### PR TITLE
Use the same python interpreter as yum when possible

### LINKED ISSUE #3143: Amazon Linux: yum package resource fails with "ImportError: No module named yum"
"Amazon Linux AMI 2014.09" servers starting failing upon attempting to install any package with chef-client. The failure looks like this for any "package" install you try:

Error executing action `install` on resource 'yum_package[tzdata]'
    ================================================================================

    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '1'
    ---- Begin output of /usr/bin/python /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30 ----
    STDOUT:
    STDERR: Traceback (most recent call last):
      File "/opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py", line 36, in <module>
        import yum
    ImportError: No module named yum
    ---- End output of /usr/bin/python /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30 ----
    Ran /usr/bin/python /opt/chef/embedded/apps/chef/lib/chef/provider/package/yum-dump.py --options --installed-provides --yum-lock-timeout 30 returned 1

    Resource Declaration:
    ---------------------
    # In /etc/chef/local-mode-cache/cache/cookbooks/timezone-ii/recipes/default.rb

     13: package value_for_platform_family(
     14:   'gentoo'  => 'timezone-data',

    Compiled Resource:
    ------------------
    # Declared in /etc/chef/local-mode-cache/cache/cookbooks/timezone-ii/recipes/default.rb:13:in `from_file'

    yum_package("tzdata") do
      action :install
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      package_name "tzdata"
      timeout 900
      flush_cache {:before=>false, :after=>false}
      declared_type :package
      cookbook_name "timezone-ii"
      recipe_name "default"
    end

After investigation, I found that the package which I update in my userdata "aws-cfn-bootstrap" was updated sometime in the last 24 hours, updating its dependencies to install python27. yum-dump.py seems to be calling /usr/bin/python, which is symlinked to python2.6 after the 2.7 install. /usr/bin/yum header was updated to "#!/usr/bin/python2.7". The problem seems to be that the 2.7 install breaks 2.6 interactivity with yum.

### PR DESCRIPTION
**Pull request by stevendanna**
We rely on a system python installation with the correct python
libraries.  To improve the chances that the python interpreter we call
has the yum libary available, we use the same interpreter as yum uses.

Fixes #3143

### CODE REVIEW

**Comment by thommay** [comment]
:+1:

File: lib/chef/provider/package/yum.rb [hunk 1]
```diff
@@ -779,6 +781,32 @@ def refresh
             @next_refresh = :none
           end
 
+          def python_bin
+            yum_executable = which("yum")
+            if yum_executable && shabang?(yum_executable)
+              extract_interpreter(yum_executable)
+            else
+              Chef::Log.warn("Yum exectuable not found or doesn't start with #!. Using default python.")
```
Comment (thommay) [inline lib/chef/provider/package/yum.rb]: "executable"

**Comment by ranjib** [comment]
:+1: 
im thinking of `/usr/bin/env python` style shebang lines, should be fine though

**Comment by stevendanna** [comment]
Can one of the Enterprise Linux maintainers take a quick look? cc @jonlives @lamont-granquist @jaymzh

**Comment by jonlives** [comment]
:+1:

**Comment by lamont-granquist** [comment]
:+1:

**Comment by jaymzh** [comment]
:+1:

**Pull request merged**